main.go

```diff
@@ -1,70 +1,133 @@
+	KeyPath           string `env:"KEY_PATH,required"`
+	RootCA            string `env:"ROOT_CA,required"`
+	PORT              int    `env:"PORT,required"`
+	RedirectHost      string `env:"REDIRECT_HOST,required"`
+	WellKnown         string `env:"WELL_KNOWN_URL,required"`
+	WellKnownURL      *url.URL
+	IssuerURL         *url.URL `env:"ISSUER_URL"`
+	AuthorizeEndpoint *url.URL `env:"AUTHORIZATION_ENDPOINT"`
+	TokenEndpoint     *url.URL `env:"MTLS_ENDPOINT_ALIASES_TOKEN_ENDPOINT"`
+	UsePyron          bool     `env:"USE_PYRON,required"`
+	ResourceURL       string   `env:"RESOURCE_URL"`
+	XSSLCertHash      string   `env:"X_SSL_CERT_HASH"`
+	InjectCertMode    bool     `env:"INJECT_CERT_MODE,required"`
+}
+
+func (c Config) NewClientConfig() acp.Config {
+	var (
+		redirectURL *url.URL
+		err         error
+	)
+
+	if redirectURL, err = url.Parse(fmt.Sprintf("http://%v:%v/callback", c.RedirectHost, c.PORT)); err != nil {
+		log.Fatalf("failed to get callback url from host %v", err)
+	}
+
+	return acp.Config{
+		ClientID:     c.ClientID,
+		RedirectURL:  redirectURL,
+		TokenURL:     c.TokenEndpoint,
+		AuthorizeURL: c.AuthorizeEndpoint,
+		IssuerURL:    c.IssuerURL,
+		CertFile:     c.CertPath,
+		KeyFile:      c.KeyPath,
+		RootCA:       c.RootCA,
+		Scopes:       []string{"openid"},
+	}
+}
+
+func LoadConfig() (config Config, err error) {
+	if err = env.Parse(&config); err != nil {
+		return config, err
+	}
+
+	if config.WellKnownURL, err = url.Parse(config.WellKnown); err != nil {
+		log.Fatalf("failed to parse wellknown as url %v", err)
+	}
+
+	config.fetchEndpointURLs()
+	usePyron = config.UsePyron
+	xsslCertHash = config.XSSLCertHash
+	resourceURL = config.ResourceURL
+	injectCertMode = config.InjectCertMode
+
+	return config, err
+}
+
+func layoutFiles() []string {
+	files, err := filepath.Glob("templates/*.html")
+	if err != nil {
+		log.Fatal(err)
+	}
+	return files
+}
 
 func main() {
 	var (
-		serverPort int
-		acpClient  acp.Client
-		err        error
+		config       Config
+		authorizeURL string
+		csrf         acp.CSRF
+		client       acp.Client
+		err          error
 	)
 
-	flag.Parse()
-	if serverPort, err = strconv.Atoi(*port); err != nil {
-		log.Fatalln(err)
+	if config, err = LoadConfig(); err != nil {
+		log.Fatalf("failed to load config %v", err)
 	}
 
-	acpOAuthConfig = acp.Config{
-		RedirectURL: fmt.Sprintf("http://%v:%v/callback", *redirectHost, serverPort),
-		ClientID:    *clientID,
-		Scopes:      []string{"openid"},
-		AuthURL:     fmt.Sprintf("%v/oauth2/authorize", *issuerURL),
-		TokenURL:    fmt.Sprintf("%v/oauth2/token", *issuerURL),
-		PKCEEnabled: *pkceEnabled,
+	if client, err = acp.New(config.NewClientConfig()); err != nil {
+		log.Fatalf("failed to get oauth client %v", err)
 	}
 
-	if acpClient, err = acp.NewClient(*serverCertPath, *certPath, *keyPath, acpOAuthConfig); err != nil {
-		log.Fatalln(err)
+	if authorizeURL, csrf, err = client.AuthorizeURL(); err != nil {
```

Comment:
This should be moved to /login endpoint logic + you need to store csrf in the cookie.